File: CarLine.DataCleanUp/Services/Cleanup/MongoUpsertBatch.cs
using CarLine.DataCleanUp.Models;
using MongoDB.Bson;
using MongoDB.Driver;

namespace CarLine.DataCleanUp.Services.Cleanup;

internal sealed class MongoUpsertBatch
{
    private readonly List<WriteModel<BsonDocument>> _batch = new();

    public int Count => _batch.Count;
    public IReadOnlyList<WriteModel<BsonDocument>> Items => _batch;

    public void Clear()
    {
        _batch.Clear();
    }

    public bool TryAddFromFullRecord(Dictionary<string, string> fullRecord, out string? url)
    {
        url = null;

        // Prepare MongoDB document (includes web display fields + core fields)
        var cleanedDoc = new BsonDocument();
        foreach (var key in DataCleanupConstants.WebDisplayFields)
            if (fullRecord.TryGetValue(key, out var val) && !string.IsNullOrWhiteSpace(val))
                cleanedDoc[key] = val;

        if (cleanedDoc.ElementCount == 0)
            return false;

        if (!fullRecord.TryGetValue("url", out url) || string.IsNullOrWhiteSpace(url))
            return false;

        // Insert-only: if url exists, Mongo unique index will reject it (E11000) and we'll ignore duplicates.
        cleanedDoc["url"] = url;
        cleanedDoc["first_seen"] = DateTime.UtcNow;
        cleanedDoc["last_seen"] = DateTime.UtcNow;
        cleanedDoc["status"] = "ACTIVE";

        _batch.Add(new InsertOneModel<BsonDocument>(cleanedDoc));
        return true;
    }
}

File: CarLine.DataCleanUp.FSharp/Infrastructure/MongoUpsertBatch.fs
namespace CarLine.DataCleanUp.FSharp

open System
open System.Collections.Generic
open MongoDB.Bson
open MongoDB.Driver

/// Accumulates MongoDB InsertOneModel entries for a single bulk-write batch.
/// Returns 'string option' from TryAdd instead of C#'s 'out string?' — more idiomatic F#.
type MongoUpsertBatch() =
    let batch = List<WriteModel<BsonDocument>>()

    member _.Count = batch.Count
    member _.Items = batch :> IReadOnlyList<WriteModel<BsonDocument>>
    member _.Clear() = batch.Clear()

    /// Returns Some url if the document was queued for insertion, None if skipped.
    member _.TryAdd(fullRecord: Dictionary<string, string>) : string option =
        let doc = BsonDocument()
        for key in DataCleanupConstants.webDisplayFields do
            match fullRecord.TryGetValue(key) with
            | true, v when not (String.IsNullOrWhiteSpace(v)) -> doc[key] <- BsonString(v)
            | _ -> ()

        if doc.ElementCount = 0 then None
        else
            match fullRecord.TryGetValue("url") with
            | true, url when not (String.IsNullOrWhiteSpace(url)) ->
                doc["url"]        <- BsonString(url)
                doc["first_seen"] <- BsonDateTime(DateTime.UtcNow)
                doc["last_seen"]  <- BsonDateTime(DateTime.UtcNow)
                doc["status"]     <- BsonString("ACTIVE")
                batch.Add(InsertOneModel<BsonDocument>(doc))
                Some url
            | _ -> None

